nemo/utils/callbacks/preemption.py

```diff
@@ -0,0 +1,93 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import signal
+import sys
+
+import torch
+from pytorch_lightning.callbacks import Callback
+
+from nemo.utils import logging
+
+
+class PreemptionCallback(Callback):
+    """
+    PreemptionCallback class creates a callback that checks for preemption during training at the end of every step.
+    Upon preemption the callback provides function to gracefully exit the training immediately and also saves the state of training (to be able to start from the 
+    same step without wasting any compute while resuming the next time).
+
+    PreemptionCallback is always enabled.
+    """
+
+    def __init__(self, checkpoint_callback, sig=signal.SIGTERM):
```

Comment:
Should be a None by default, and if None then self.sig = [signal.SIGTERM]. That way BCP can pass other signals.